{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tryptophan, four kynurenine-pathway metabolites (kynurenine, kynurenate, xanthurenate and quinolinate)\n",
      "                   and indolelactate were positively associated with T2D risk, while indolepropionate was inversely \n",
      "                   associated with T2D risk. We identified multiple host genetic variants, dietary factors, gut bacteria \n",
      "                   and their potential interplay associated with these T2D-relaetd metabolites. Intakes of fibre-rich foods, \n",
      "                   but not protein/tryptophan-rich foods, were the dietary factors most strongly associated with tryptophan \n",
      "                   metabolites. The fibre-indolepropionate association was partially explained by indolepropionate-associated \n",
      "                   gut bacteria, mostly fibre-using Firmicutes. We identified a novel association between a host functional \n",
      "                   LCT variant (determining lactase persistence) and serum indolepropionate, which might be related to a host \n",
      "                   gene-diet interaction on gut Bifidobacterium, a probiotic bacterium significantly associated with \n",
      "                   indolepropionate independent of other fibre-related bacteria. Higher milk intake was associated with higher \n",
      "                   levels of gut Bifidobacterium and serum indolepropionate only among genetically lactase non-persistent \n",
      "                   individuals.\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "from promptify import OpenAI\n",
    "from promptify import Prompter\n",
    "from pprint import pprint\n",
    "from IPython.display import Markdown, display, HTML\n",
    "\n",
    "# Define the API key for the OpenAI model\n",
    "api_key  = \"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\"\n",
    "\n",
    "\n",
    "# Create an instance of the OpenAI model, Currently supporting Openai's all model, In future adding more generative models from Hugginface and other platforms\n",
    "model = OpenAI(api_key, model=\"gpt-3.5-turbo\")\n",
    "nlp_prompter = Prompter(model)\n",
    "\n",
    "# Example sentence for demonstration\n",
    "sentence     =  \"\"\"Tryptophan, four kynurenine-pathway metabolites (kynurenine, kynurenate, xanthurenate and quinolinate)\n",
    "                   and indolelactate were positively associated with T2D risk, while indolepropionate was inversely \n",
    "                   associated with T2D risk. We identified multiple host genetic variants, dietary factors, gut bacteria \n",
    "                   and their potential interplay associated with these T2D-relaetd metabolites. Intakes of fibre-rich foods, \n",
    "                   but not protein/tryptophan-rich foods, were the dietary factors most strongly associated with tryptophan \n",
    "                   metabolites. The fibre-indolepropionate association was partially explained by indolepropionate-associated \n",
    "                   gut bacteria, mostly fibre-using Firmicutes. We identified a novel association between a host functional \n",
    "                   LCT variant (determining lactase persistence) and serum indolepropionate, which might be related to a host \n",
    "                   gene-diet interaction on gut Bifidobacterium, a probiotic bacterium significantly associated with \n",
    "                   indolepropionate independent of other fibre-related bacteria. Higher milk intake was associated with higher \n",
    "                   levels of gut Bifidobacterium and serum indolepropionate only among genetically lactase non-persistent \n",
    "                   individuals.\"\"\"\n",
    "\n",
    "print(sentence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " [['Tryptophan', 'positively associated with', 'T2D risk'],\n",
      "        ['four kynurenine-pathway metabolites', 'positively associated with', 'T2D risk'],\n",
      "        ['kynurenine', 'positively associated with', 'T2D risk'],\n",
      "        ['kynurenate', 'positively associated with', 'T2D risk'],\n",
      "        ['xanthurenate', 'positively associated with', 'T2D risk'],\n",
      "        ['quinolinate', 'positively associated with', 'T2D risk'],\n",
      "        ['indolelactate', 'positively associated with', 'T2D risk'],\n",
      "        ['indolepropionate', 'inversely associated with', 'T2D risk'],\n",
      "        ['host genetic variants', 'associated with', 'T2D-related metabolites'],\n",
      "        ['dietary factors', 'associated with', 'T2D-related metabolites'],\n",
      "        ['gut bacteria', 'associated with', 'T2D-related metabolites'],\n",
      "        ['fibre-rich foods', 'associated with', 'tryptophan metabolites'],\n",
      "        ['protein/tryptophan-rich foods', 'not associated with', 'tryptophan metabolites'],\n",
      "        ['fibre-indolepropionate association', 'partially explained by', 'indolepropionate-associated gut bacteria'],\n",
      "        ['Firmicutes', 'associated with', 'indolepropionate-associated gut bacteria'],\n",
      "        ['host functional LCT variant', 'associated with', 'serum indolepropionate'],\n",
      "        ['host gene-diet interaction', 'associated with', 'gut Bifidobacterium'],\n",
      "        ['probiotic bacterium', 'significantly associated with', 'indolepropionate'],\n",
      "        ['milk intake', 'associated with', 'higher levels of gut Bifidobacterium'],\n",
      "        ['milk intake', 'associated with', 'serum indolepropionate'],\n",
      "        ['genetically lactase non-persistent individuals', 'associated with', 'higher milk intake']\n",
      "        ]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<h4>Sentence</h4>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Tryptophan, four kynurenine-pathway metabolites (kynurenine, kynurenate, xanthurenate and quinolinate)\n",
      "                   and indolelactate were positively associated with T2D risk, while indolepropionate was inversely \n",
      "                   associated with T2D risk. We identified multiple host genetic variants, dietary factors, gut bacteria \n",
      "                   and their potential interplay associated with these T2D-relaetd metabolites. Intakes of fibre-rich foods, \n",
      "                   but not protein/tryptophan-rich foods, were the dietary factors most strongly associated with tryptophan \n",
      "                   metabolites. The fibre-indolepropionate association was partially explained by indolepropionate-associated \n",
      "                   gut bacteria, mostly fibre-using Firmicutes. We identified a novel association between a host functional \n",
      "                   LCT variant (determining lactase persistence) and serum indolepropionate, which might be related to a host \n",
      "                   gene-diet interaction on gut Bifidobacterium, a probiotic bacterium significantly associated with \n",
      "                   indolepropionate independent of other fibre-related bacteria. Higher milk intake was associated with higher \n",
      "                   levels of gut Bifidobacterium and serum indolepropionate only among genetically lactase non-persistent \n",
      "                   individuals.\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<h4>Output</h4>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "[['Tryptophan', 'positively associated with', 'T2D risk'],\n",
      " ['four kynurenine-pathway metabolites',\n",
      "  'positively associated with',\n",
      "  'T2D risk'],\n",
      " ['kynurenine', 'positively associated with', 'T2D risk'],\n",
      " ['kynurenate', 'positively associated with', 'T2D risk'],\n",
      " ['xanthurenate', 'positively associated with', 'T2D risk'],\n",
      " ['quinolinate', 'positively associated with', 'T2D risk'],\n",
      " ['indolelactate', 'positively associated with', 'T2D risk'],\n",
      " ['indolepropionate', 'inversely associated with', 'T2D risk'],\n",
      " ['host genetic variants', 'associated with', 'T2D-related metabolites'],\n",
      " ['dietary factors', 'associated with', 'T2D-related metabolites'],\n",
      " ['gut bacteria', 'associated with', 'T2D-related metabolites'],\n",
      " ['fibre-rich foods', 'associated with', 'tryptophan metabolites'],\n",
      " ['protein/tryptophan-rich foods',\n",
      "  'not associated with',\n",
      "  'tryptophan metabolites'],\n",
      " ['fibre-indolepropionate association',\n",
      "  'partially explained by',\n",
      "  'indolepropionate-associated gut bacteria'],\n",
      " ['Firmicutes', 'associated with', 'indolepropionate-associated gut bacteria'],\n",
      " ['host functional LCT variant', 'associated with', 'serum indolepropionate'],\n",
      " ['host gene-diet interaction', 'associated with', 'gut Bifidobacterium'],\n",
      " ['probiotic bacterium', 'significantly associated with', 'indolepropionate'],\n",
      " ['milk intake', 'associated with', 'higher levels of gut Bifidobacterium'],\n",
      " ['milk intake', 'associated with', 'serum indolepropionate'],\n",
      " ['genetically lactase non-persistent individuals',\n",
      "  'associated with',\n",
      "  'higher milk intake']]\n"
     ]
    }
   ],
   "source": [
    "result       = nlp_prompter.fit('relation_extraction.jinja', \n",
    "                                 domain = 'clinical',\n",
    "                                 text_input  = sentence)\n",
    "\n",
    "print(result['text'])\n",
    "\n",
    "# Output\n",
    "\n",
    "display(HTML('<h4>Sentence</h4>'))\n",
    "print('\\n')\n",
    "print(sentence)\n",
    "print('\\n')\n",
    "display(HTML('<h4>Output</h4>'))\n",
    "print(\"\\n\")\n",
    "pprint(eval(result['text']))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load examples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Got 2 examples.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "\n",
    "relation_examples = json.load(open(\"promptify/data/text2text/en/relation_extraction.json\",'r'))\n",
    "print(\"Got\", len(relation_examples), \"examples.\")\n",
    "\n",
    "prompt_examples = []\n",
    "for sample in relation_examples:\n",
    "    prompt_examples.append((sample['text'], sample['labels']))\n",
    "print()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test BERN2 availability"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'_id': '30429607', 'annotations': [{'id': ['mesh:D009369'], 'is_neural_normalized': False, 'mention': 'tumour', 'obj': 'disease', 'prob': 0.9999922513961792, 'span': {'begin': 20, 'end': 26}}, {'id': ['mesh:D001120'], 'is_neural_normalized': False, 'mention': 'arginine', 'obj': 'drug', 'prob': 0.9819279909133911, 'span': {'begin': 54, 'end': 62}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'lysosomes', 'obj': 'cell_type', 'prob': 0.4410809576511383, 'span': {'begin': 129, 'end': 138}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'essential autophagy gene', 'obj': 'DNA', 'prob': 0.8187922835350037, 'span': {'begin': 280, 'end': 304}}, {'id': ['NCBIGene:8408'], 'is_neural_normalized': True, 'mention': 'autophagy gene', 'obj': 'gene', 'prob': 0.695938229560852, 'span': {'begin': 290, 'end': 304}}, {'id': ['NCBIGene:10533'], 'is_neural_normalized': False, 'mention': 'Atg7', 'obj': 'gene', 'prob': 0.6825664043426514, 'span': {'begin': 305, 'end': 309}}, {'id': ['NCBITaxon:10095'], 'is_neural_normalized': False, 'mention': 'mice', 'obj': 'species', 'prob': 0.9752292633056641, 'span': {'begin': 319, 'end': 323}}, {'id': ['mesh:D013217'], 'is_neural_normalized': True, 'mention': 'starvation intolerance', 'obj': 'disease', 'prob': 0.9323878884315491, 'span': {'begin': 377, 'end': 399}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'Cancer cells', 'obj': 'cell_type', 'prob': 0.8247388601303101, 'span': {'begin': 477, 'end': 489}}, {'id': ['mesh:D009369'], 'is_neural_normalized': False, 'mention': 'Cancer', 'obj': 'disease', 'prob': 0.9979585409164429, 'span': {'begin': 477, 'end': 483}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'essential autophagy genes', 'obj': 'DNA', 'prob': 0.7361432909965515, 'span': {'begin': 531, 'end': 556}}, {'id': ['NCBIGene:8408'], 'is_neural_normalized': True, 'mention': 'autophagy', 'obj': 'gene', 'prob': 0.8590404391288757, 'span': {'begin': 541, 'end': 550}}, {'id': ['mesh:D009369'], 'is_neural_normalized': True, 'mention': 'spontaneous tumours', 'obj': 'disease', 'prob': 0.9996308088302612, 'span': {'begin': 623, 'end': 642}}, {'id': ['mesh:D009369'], 'is_neural_normalized': True, 'mention': 'autochthonous cancer', 'obj': 'disease', 'prob': 0.9999895691871643, 'span': {'begin': 656, 'end': 676}}, {'id': ['NCBIGene:10533'], 'is_neural_normalized': False, 'mention': 'Atg7', 'obj': 'gene', 'prob': 0.9961122870445251, 'span': {'begin': 711, 'end': 715}}, {'id': ['NCBIGene:23192'], 'is_neural_normalized': True, 'mention': 'dominant-negative ATG4b', 'obj': 'gene', 'prob': 0.9988656044006348, 'span': {'begin': 751, 'end': 774}}, {'id': ['NCBITaxon:10095'], 'is_neural_normalized': False, 'mention': 'mice', 'obj': 'species', 'prob': 0.9348741769790649, 'span': {'begin': 778, 'end': 782}}, {'id': ['mesh:D009369'], 'is_neural_normalized': True, 'mention': 'KRAS-driven cancers', 'obj': 'disease', 'prob': 0.9999643564224243, 'span': {'begin': 813, 'end': 832}}, {'id': ['NCBIGene:3845'], 'is_neural_normalized': False, 'mention': 'KRAS', 'obj': 'gene', 'prob': 0.9355710744857788, 'span': {'begin': 813, 'end': 817}}, {'id': ['mesh:D009369'], 'is_neural_normalized': False, 'mention': 'tumour', 'obj': 'disease', 'prob': 0.9999947547912598, 'span': {'begin': 843, 'end': 849}}, {'id': ['mesh:D009369'], 'is_neural_normalized': False, 'mention': 'tumour', 'obj': 'disease', 'prob': 0.9999812841415405, 'span': {'begin': 923, 'end': 929}}, {'id': ['NCBIGene:10533'], 'is_neural_normalized': False, 'mention': 'Atg7', 'obj': 'gene', 'prob': 0.9903233051300049, 'span': {'begin': 987, 'end': 991}}, {'id': ['mesh:D009378'], 'is_neural_normalized': True, 'mention': 'multiple allografted tumours', 'obj': 'disease', 'prob': 0.8184354901313782, 'span': {'begin': 1014, 'end': 1042}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'tumour lines', 'obj': 'cell_line', 'prob': 0.6105258464813232, 'span': {'begin': 1061, 'end': 1073}}, {'id': ['mesh:D009369'], 'is_neural_normalized': False, 'mention': 'tumour', 'obj': 'disease', 'prob': 0.9999525547027588, 'span': {'begin': 1061, 'end': 1067}}, {'id': ['mesh:D001120'], 'is_neural_normalized': False, 'mention': 'arginine', 'obj': 'drug', 'prob': 0.9946534633636475, 'span': {'begin': 1192, 'end': 1200}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'tumour cell lines', 'obj': 'cell_line', 'prob': 0.8817534446716309, 'span': {'begin': 1220, 'end': 1237}}, {'id': ['mesh:D009369'], 'is_neural_normalized': False, 'mention': 'tumour', 'obj': 'disease', 'prob': 0.9997414946556091, 'span': {'begin': 1220, 'end': 1226}}, {'id': ['mesh:D001120'], 'is_neural_normalized': False, 'mention': 'arginine', 'obj': 'drug', 'prob': 0.9986180067062378, 'span': {'begin': 1243, 'end': 1251}}, {'id': ['mesh:D001125'], 'is_neural_normalized': True, 'mention': 'argininosuccinate', 'obj': 'drug', 'prob': 0.9984564781188965, 'span': {'begin': 1309, 'end': 1326}}, {'id': ['NCBIGene:445'], 'is_neural_normalized': False, 'mention': 'argininosuccinate synthase 1', 'obj': 'gene', 'prob': 0.9427773952484131, 'span': {'begin': 1309, 'end': 1337}}, {'id': ['mesh:D001120'], 'is_neural_normalized': False, 'mention': 'arginine', 'obj': 'drug', 'prob': 0.9970167875289917, 'span': {'begin': 1379, 'end': 1387}}, {'id': ['NCBIGene:383'], 'is_neural_normalized': True, 'mention': 'arginase I', 'obj': 'gene', 'prob': 0.8367504477500916, 'span': {'begin': 1405, 'end': 1415}}, {'id': ['NCBIGene:383'], 'is_neural_normalized': False, 'mention': 'ARG1', 'obj': 'gene', 'prob': 0.9970721006393433, 'span': {'begin': 1417, 'end': 1421}}, {'id': ['NCBIGene:10533'], 'is_neural_normalized': False, 'mention': 'Atg7', 'obj': 'gene', 'prob': 0.9977797865867615, 'span': {'begin': 1445, 'end': 1449}}, {'id': ['mesh:D001120'], 'is_neural_normalized': False, 'mention': 'arginine', 'obj': 'drug', 'prob': 0.9989784955978394, 'span': {'begin': 1479, 'end': 1487}}, {'id': ['mesh:D001120'], 'is_neural_normalized': False, 'mention': 'arginine', 'obj': 'drug', 'prob': 0.9990563988685608, 'span': {'begin': 1530, 'end': 1538}}, {'id': ['mesh:D009952'], 'is_neural_normalized': False, 'mention': 'ornithine', 'obj': 'drug', 'prob': 0.9229053854942322, 'span': {'begin': 1555, 'end': 1564}}, {'id': ['NCBIGene:383'], 'is_neural_normalized': False, 'mention': 'ARG1', 'obj': 'gene', 'prob': 0.9765568971633911, 'span': {'begin': 1566, 'end': 1570}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'hepatocytes', 'obj': 'cell_type', 'prob': 0.5102328658103943, 'span': {'begin': 1636, 'end': 1647}}, {'id': ['NCBIGene:10533'], 'is_neural_normalized': False, 'mention': 'Atg7', 'obj': 'gene', 'prob': 0.9929279685020447, 'span': {'begin': 1697, 'end': 1701}}, {'id': ['NCBIGene:383'], 'is_neural_normalized': False, 'mention': 'ARG1', 'obj': 'gene', 'prob': 0.9825230836868286, 'span': {'begin': 1723, 'end': 1727}}, {'id': ['mesh:D001120'], 'is_neural_normalized': False, 'mention': 'arginine', 'obj': 'drug', 'prob': 0.9920065999031067, 'span': {'begin': 1752, 'end': 1760}}, {'id': ['mesh:D009369'], 'is_neural_normalized': False, 'mention': 'tumour', 'obj': 'disease', 'prob': 0.9999942779541016, 'span': {'begin': 1765, 'end': 1771}}, {'id': ['NCBIGene:9474'], 'is_neural_normalized': False, 'mention': 'Atg5', 'obj': 'gene', 'prob': 0.9938210844993591, 'span': {'begin': 1792, 'end': 1796}}, {'id': ['mesh:D001120'], 'is_neural_normalized': False, 'mention': 'arginine', 'obj': 'drug', 'prob': 0.9871364831924438, 'span': {'begin': 1853, 'end': 1861}}, {'id': ['mesh:D063646'], 'is_neural_normalized': False, 'mention': 'tumorigenesis', 'obj': 'disease', 'prob': 0.8955700993537903, 'span': {'begin': 1877, 'end': 1890}}, {'id': ['NCBIGene:10533'], 'is_neural_normalized': False, 'mention': 'Atg7', 'obj': 'gene', 'prob': 0.9864484667778015, 'span': {'begin': 1992, 'end': 1996}}, {'id': ['NCBIGene:10533'], 'is_neural_normalized': False, 'mention': 'Atg7', 'obj': 'gene', 'prob': 0.997479259967804, 'span': {'begin': 2025, 'end': 2029}}, {'id': ['mesh:D001120'], 'is_neural_normalized': False, 'mention': 'arginine', 'obj': 'drug', 'prob': 0.9982208609580994, 'span': {'begin': 2051, 'end': 2059}}, {'id': ['mesh:D001120'], 'is_neural_normalized': False, 'mention': 'arginine', 'obj': 'drug', 'prob': 0.9987165927886963, 'span': {'begin': 2101, 'end': 2109}}, {'id': ['mesh:D009369'], 'is_neural_normalized': False, 'mention': 'tumour', 'obj': 'disease', 'prob': 0.9999785423278809, 'span': {'begin': 2114, 'end': 2120}}, {'id': ['NCBIGene:383'], 'is_neural_normalized': False, 'mention': 'ARG1', 'obj': 'gene', 'prob': 0.9866706728935242, 'span': {'begin': 2191, 'end': 2195}}, {'id': ['mesh:D001120'], 'is_neural_normalized': False, 'mention': 'arginine', 'obj': 'drug', 'prob': 0.9935520887374878, 'span': {'begin': 2246, 'end': 2254}}, {'id': ['mesh:D009369'], 'is_neural_normalized': False, 'mention': 'tumour', 'obj': 'disease', 'prob': 0.9999873638153076, 'span': {'begin': 2279, 'end': 2285}}, {'id': ['mesh:D009369'], 'is_neural_normalized': False, 'mention': 'cancer', 'obj': 'disease', 'prob': 0.9976287484169006, 'span': {'begin': 2339, 'end': 2345}}], 'pmid': '30429607', 'text': 'Autophagy maintains tumour growth through circulating arginine. Autophagy captures intracellular components and delivers them to lysosomes, where they are degraded and recycled to sustain metabolism and to enable survival during starvation(1-5). Acute, whole-body deletion of the essential autophagy gene Atg7 in adult mice causes a systemic metabolic defect that manifests as starvation intolerance and gradual loss of white adipose tissue, liver glycogen and muscle mass(1). Cancer cells also benefit from autophagy. Deletion of essential autophagy genes impairs the metabolism, proliferation, survival and malignancy of spontaneous tumours in models of autochthonous cancer(6,7). Acute, systemic deletion of Atg7 or acute, systemic expression of a dominant-negative ATG4b in mice induces greater regression of KRAS-driven cancers than does tumour-specific autophagy deletion, which suggests that host autophagy promotes tumour growth(1,8). Here we show that host-specific deletion of Atg7 impairs the growth of multiple allografted tumours, although not all tumour lines were sensitive to host autophagy status. Loss of autophagy in the host was associated with a reduction in circulating arginine, and the sensitive tumour cell lines were arginine auxotrophs owing to the lack of expression of the enzyme argininosuccinate synthase 1. Serum proteomic analysis identified the arginine-degrading enzyme arginase I (ARG1) in the circulation of Atg7-deficient hosts, and in vivo arginine metabolic tracing demonstrated that serum arginine was degraded to ornithine. ARG1 is predominantly expressed in the liver and can be released from hepatocytes into the circulation. Liver-specific deletion of Atg7 produced circulating ARG1, and reduced both serum arginine and tumour growth. Deletion of Atg5 in the host similarly regulated [corrected] circulating arginine and suppressed tumorigenesis, which demonstrates that this phenotype is specific to autophagy function rather than to deletion of Atg7. Dietary supplementation of Atg7-deficient hosts with arginine partially restored levels of circulating arginine and tumour growth. Thus, defective autophagy in the host leads to the release of ARG1 from the liver and the degradation of circulating arginine, which is essential for tumour growth; this identifies a metabolic vulnerability of cancer.', 'timestamp': 'Sun Apr 30 17:06:59 +0000 2023'}, {'_id': '29446767', 'annotations': [{'id': ['mesh:C567763'], 'is_neural_normalized': False, 'mention': 'CLAPO syndrome', 'obj': 'disease', 'prob': 0.9999992847442627, 'span': {'begin': 0, 'end': 14}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'PIK3CA', 'obj': 'DNA', 'prob': 0.7425963282585144, 'span': {'begin': 53, 'end': 59}}, {'id': ['NCBIGene:5290'], 'is_neural_normalized': True, 'mention': 'PIK3CA mutations', 'obj': 'gene', 'prob': 0.7837176322937012, 'span': {'begin': 53, 'end': 69}}, {'id': ['mesh:C567763'], 'is_neural_normalized': False, 'mention': 'CLAPO syndrome', 'obj': 'disease', 'prob': 0.9999993443489075, 'span': {'begin': 83, 'end': 97}}, {'id': ['mesh:D014652'], 'is_neural_normalized': False, 'mention': 'vascular disorder', 'obj': 'disease', 'prob': 0.9999779462814331, 'span': {'begin': 108, 'end': 125}}, {'id': ['mesh:C567763'], 'is_neural_normalized': True, 'mention': 'capillary malformation of the lower lip', 'obj': 'disease', 'prob': 0.9999973177909851, 'span': {'begin': 143, 'end': 182}}, {'id': ['mesh:D044148'], 'is_neural_normalized': True, 'mention': 'lymphatic malformation', 'obj': 'disease', 'prob': 0.999998927116394, 'span': {'begin': 184, 'end': 206}}, {'id': ['mesh:D000014'], 'is_neural_normalized': True, 'mention': 'asymmetry', 'obj': 'disease', 'prob': 0.9982122182846069, 'span': {'begin': 241, 'end': 250}}, {'id': ['mesh:D015576'], 'is_neural_normalized': True, 'mention': 'overgrowth', 'obj': 'disease', 'prob': 0.882811963558197, 'span': {'begin': 276, 'end': 286}}, {'id': ['mesh:C567763'], 'is_neural_normalized': False, 'mention': 'CLAPO', 'obj': 'disease', 'prob': 0.9999896287918091, 'span': {'begin': 423, 'end': 428}}, {'id': ['NCBITaxon:9606'], 'is_neural_normalized': False, 'mention': 'patients', 'obj': 'species', 'prob': 0.9683281183242798, 'span': {'begin': 525, 'end': 533}}, {'id': ['mesh:C567763'], 'is_neural_normalized': False, 'mention': 'CLAPO', 'obj': 'disease', 'prob': 0.9999940395355225, 'span': {'begin': 539, 'end': 544}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'DNA', 'obj': 'DNA', 'prob': 0.6079210638999939, 'span': {'begin': 561, 'end': 564}}, {'id': ['NCBITaxon:9606'], 'is_neural_normalized': False, 'mention': 'patients', 'obj': 'species', 'prob': 0.9668839573860168, 'span': {'begin': 593, 'end': 601}}, {'id': ['NCBIGene:5290'], 'is_neural_normalized': False, 'mention': 'PIK3CA gene', 'obj': 'gene', 'prob': 0.9959104061126709, 'span': {'begin': 698, 'end': 709}}, {'id': ['NCBITaxon:9606'], 'is_neural_normalized': False, 'mention': 'patients', 'obj': 'species', 'prob': 0.9671095013618469, 'span': {'begin': 746, 'end': 754}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'NM_006218', 'obj': 'RNA', 'prob': 0.675126314163208, 'span': {'begin': 785, 'end': 794}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'c.', 'obj': 'RNA', 'prob': 0.5695823431015015, 'span': {'begin': 797, 'end': 799}}, {'id': ['mesh:D002658'], 'is_neural_normalized': False, 'mention': 'developmental disorders', 'obj': 'disease', 'prob': 0.9999232888221741, 'span': {'begin': 859, 'end': 882}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'PIK3CA', 'obj': 'DNA', 'prob': 0.8881197571754456, 'span': {'begin': 962, 'end': 968}}, {'id': ['NCBIGene:5290'], 'is_neural_normalized': True, 'mention': 'PIK3CA mutations', 'obj': 'gene', 'prob': 0.7521552443504333, 'span': {'begin': 962, 'end': 978}}, {'id': ['NCBITaxon:9606'], 'is_neural_normalized': False, 'mention': 'patients', 'obj': 'species', 'prob': 0.9286090135574341, 'span': {'begin': 982, 'end': 990}}, {'id': ['mesh:C567763'], 'is_neural_normalized': False, 'mention': 'CLAPO', 'obj': 'disease', 'prob': 0.9999841451644897, 'span': {'begin': 996, 'end': 1001}}], 'pmid': '29446767', 'text': 'CLAPO syndrome: identification of somatic activating PIK3CA mutations and PURPOSE: CLAPO syndrome is a rare vascular disorder characterized by capillary malformation of the lower lip, lymphatic malformation predominant on the face and neck, asymmetry, and partial/generalized overgrowth. Here we tested the hypothesis that, although the genetic cause is not known, the tissue distribution of the clinical manifestations in CLAPO seems to follow a pattern of somatic mosaicism. METHODS: We clinically evaluated a cohort of 13 patients with CLAPO and screened 20 DNA blood/tissue samples from 9 patients using high-throughput, deep sequencing. RESULTS: We identified five activating mutations in the PIK3CA gene in affected tissues from 6 of the 9 patients studied; one of the variants (NM_006218.2:c.248T&gt;C; p.Phe83Ser) has not been previously described in developmental disorders. CONCLUSION: We describe for the first time the presence of somatic activating PIK3CA mutations in patients with CLAPO. We also report an update of the phenotype and natural history of the syndrome.', 'timestamp': 'Sun Apr 30 17:07:00 +0000 2023'}]\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "\n",
    "def query_pmid(pmids, url=\"http://172.16.100.118:8888/pubmed\"):\n",
    "    return requests.get(url + \"/\" + \",\".join(pmids)).json()\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    pmids = [\"30429607\", \"29446767\"]\n",
    "    print(query_pmid(pmids))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/hanhu/code-repository/RevPromptify/examples'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "os.path.abspath(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "ename": "TemplateNotFound",
     "evalue": "/Users/hanhu/code-repository/RevPromptify/promptify/prompts/text2text/ner/ner_openai.jinja",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTemplateNotFound\u001b[0m                          Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[13], line 20\u001b[0m\n\u001b[1;32m      1\u001b[0m sentence2 \u001b[39m=\u001b[39m \u001b[39m\"\"\"\u001b[39m\u001b[39mMetformin is a common drug for the management of type 2 diabetes mellitus; however, \u001b[39m\n\u001b[1;32m      2\u001b[0m \u001b[39m                it causes various adverse gastrointestinal effects, especially after prolonged treatment.\u001b[39m\n\u001b[1;32m      3\u001b[0m \u001b[39m                It is thus of interest to identify an adjuvant treatment that synergizes with the efficacy \u001b[39m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[39m                      modulating the gut microbiome and, subsequently, SCFA and bile acid metabolism. Our findings \u001b[39m\n\u001b[1;32m     18\u001b[0m \u001b[39m                      support that cotreatment with probiotics and metformin is beneficial to patients with type 2 diabetes mellitus.\u001b[39m\u001b[39m\"\"\"\u001b[39m\n\u001b[0;32m---> 20\u001b[0m result       \u001b[39m=\u001b[39m nlp_prompter\u001b[39m.\u001b[39;49mfit(\u001b[39m'\u001b[39;49m\u001b[39m/Users/hanhu/code-repository/RevPromptify/promptify/prompts/text2text/ner/ner_openai.jinja\u001b[39;49m\u001b[39m'\u001b[39;49m, \n\u001b[1;32m     21\u001b[0m                                  domain \u001b[39m=\u001b[39;49m \u001b[39m'\u001b[39;49m\u001b[39mbiomedical\u001b[39;49m\u001b[39m'\u001b[39;49m,\n\u001b[1;32m     22\u001b[0m                                  \u001b[39m#examples = [prompt_examples[0]],\u001b[39;49;00m\n\u001b[1;32m     23\u001b[0m                                  text_input  \u001b[39m=\u001b[39;49m sentence2,\n\u001b[1;32m     24\u001b[0m                                  labels \u001b[39m=\u001b[39;49m [\u001b[39m\"\u001b[39;49m\u001b[39mGene/Protein\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39m\"\u001b[39;49m\u001b[39mDisease\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39m\"\u001b[39;49m\u001b[39mDietary supplement\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39m\"\u001b[39;49m\u001b[39mDrug/Chemical\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39m\"\u001b[39;49m\u001b[39mSpecies\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39m\"\u001b[39;49m\u001b[39mMutation\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39m\"\u001b[39;49m\u001b[39mCell Line\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39m\"\u001b[39;49m\u001b[39mCell Type\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39m\"\u001b[39;49m\u001b[39mDNA\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39m\"\u001b[39;49m\u001b[39mRNA\u001b[39;49m\u001b[39m\"\u001b[39;49m])\n\u001b[1;32m     25\u001b[0m \u001b[39m# print(result)\u001b[39;00m\n\u001b[1;32m     26\u001b[0m \u001b[39mprint\u001b[39m(result[\u001b[39m'\u001b[39m\u001b[39mtext\u001b[39m\u001b[39m'\u001b[39m])\n",
      "File \u001b[0;32m~/miniconda/envs/promptify/lib/python3.9/site-packages/promptify-0.1.4-py3.9.egg/promptify/prompts/nlp/prompter.py:130\u001b[0m, in \u001b[0;36mPrompter.fit\u001b[0;34m(self, template_name, **kwargs)\u001b[0m\n\u001b[1;32m    113\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mfit\u001b[39m(\u001b[39mself\u001b[39m, template_name: \u001b[39mstr\u001b[39m, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs):\n\u001b[1;32m    114\u001b[0m \u001b[39m    \u001b[39m\u001b[39m\"\"\"\u001b[39;00m\n\u001b[1;32m    115\u001b[0m \u001b[39m    Train the model on a generated prompt.\u001b[39;00m\n\u001b[1;32m    116\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    128\u001b[0m \n\u001b[1;32m    129\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 130\u001b[0m     prompt_variables \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mget_template_variables(template_name)\n\u001b[1;32m    131\u001b[0m     prompt_kwargs \u001b[39m=\u001b[39m {}\n\u001b[1;32m    132\u001b[0m     model_kwargs \u001b[39m=\u001b[39m {}\n",
      "File \u001b[0;32m~/miniconda/envs/promptify/lib/python3.9/site-packages/promptify-0.1.4-py3.9.egg/promptify/prompts/nlp/prompter.py:75\u001b[0m, in \u001b[0;36mPrompter.get_template_variables\u001b[0;34m(self, template_name)\u001b[0m\n\u001b[1;32m     60\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mget_template_variables\u001b[39m(\u001b[39mself\u001b[39m, template_name: \u001b[39mstr\u001b[39m) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m List[\u001b[39mstr\u001b[39m]:\n\u001b[1;32m     61\u001b[0m \u001b[39m    \u001b[39m\u001b[39m\"\"\"\u001b[39;00m\n\u001b[1;32m     62\u001b[0m \u001b[39m    Return a list of undeclared variables in the specified template.\u001b[39;00m\n\u001b[1;32m     63\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     73\u001b[0m \n\u001b[1;32m     74\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[0;32m---> 75\u001b[0m     template_source \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49menvironment\u001b[39m.\u001b[39;49mloader\u001b[39m.\u001b[39;49mget_source(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49menvironment, template_name)\n\u001b[1;32m     76\u001b[0m     parsed_content \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39menvironment\u001b[39m.\u001b[39mparse(template_source)\n\u001b[1;32m     77\u001b[0m     undeclared_variables \u001b[39m=\u001b[39m meta\u001b[39m.\u001b[39mfind_undeclared_variables(parsed_content)\n",
      "File \u001b[0;32m~/miniconda/envs/promptify/lib/python3.9/site-packages/jinja2/loaders.py:197\u001b[0m, in \u001b[0;36mFileSystemLoader.get_source\u001b[0;34m(self, environment, template)\u001b[0m\n\u001b[1;32m    194\u001b[0m             \u001b[39mreturn\u001b[39;00m \u001b[39mFalse\u001b[39;00m\n\u001b[1;32m    196\u001b[0m     \u001b[39mreturn\u001b[39;00m contents, filename, uptodate\n\u001b[0;32m--> 197\u001b[0m \u001b[39mraise\u001b[39;00m TemplateNotFound(template)\n",
      "\u001b[0;31mTemplateNotFound\u001b[0m: /Users/hanhu/code-repository/RevPromptify/promptify/prompts/text2text/ner/ner_openai.jinja"
     ]
    }
   ],
   "source": [
    "sentence2 = \"\"\"Metformin is a common drug for the management of type 2 diabetes mellitus; however, \n",
    "                it causes various adverse gastrointestinal effects, especially after prolonged treatment.\n",
    "                It is thus of interest to identify an adjuvant treatment that synergizes with the efficacy \n",
    "                of metformin while mitigating its adverse effects. Since previous evidence supports that \n",
    "                the gut microbiota is a target of metformin, this study investigated the beneficial effect \n",
    "                and mechanism of the coadministration of probiotics with metformin in the management of \n",
    "                type 2 diabetes mellitus by conducting a 3-month randomized, double-blind, placebo-controlled\n",
    "                clinical trial (n = 27 and 21 in the probiotic and placebo groups, respectively, \n",
    "                who completed the trial). Clinical results showed that the coadministration of probiotics with \n",
    "                metformin significantly reduced glycated hemoglobin compared with metformin taken alone (P < 0.05). \n",
    "                Metagenomic and metabolomic analyses showed that the coadministration of probiotics increased \n",
    "                the abundance of gut short-chain fatty acid (SCFA)-producing bacteria and bile acids.\n",
    "                      Significantly or marginally more bile acids and related metabolites were detected in the \n",
    "                      probiotic group than in the placebo group postintervention. Taken together, the results of \n",
    "                      our study showed that the coadministration of probiotics with metformin synergized with the \n",
    "                      hypoglycemic effect in patients with type 2 diabetes mellitus, which was likely through \n",
    "                      modulating the gut microbiome and, subsequently, SCFA and bile acid metabolism. Our findings \n",
    "                      support that cotreatment with probiotics and metformin is beneficial to patients with type 2 diabetes mellitus.\"\"\"\n",
    "\n",
    "result       = nlp_prompter.fit('/Users/hanhu/code-repository/RevPromptify/promptify/prompts/text2text/ner/ner_openai.jinja', \n",
    "                                 domain = 'biomedical',\n",
    "                                 #examples = [prompt_examples[0]],\n",
    "                                 text_input  = sentence2,\n",
    "                                 labels = [\"Gene/Protein\", \"Disease\", \"Dietary supplement\", \"Drug/Chemical\", \"Species\", \"Mutation\", \"Cell Line\", \"Cell Type\", \"DNA\", \"RNA\"])\n",
    "# print(result)\n",
    "print(result['text'])\n",
    "\n",
    "# Output\n",
    "\n",
    "#display(HTML('<h4>Sentence</h4>'))\n",
    "#print('\\n')\n",
    "#print(sentence)\n",
    "#print('\\n')\n",
    "#display(HTML('<h4>Output</h4>'))\n",
    "#print(\"\\n\")\n",
    "#pprint(eval(result['text']))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "promptify",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "829f9a5b4f93d769012a53136cedad0098e828d5ff1b31a47178ee9f5e17947a"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
